File: src/compare_hcp.py
import os
import sys
import argparse
import re

# Automatically add the current directory (src) to the Python path
sys.path.insert(0, os.path.abspath(os.path.dirname(__file__)))


def get_files_from_folder(folder_path):
    def extract_numbers(filename):
        numbers = re.findall(r'\d+', filename)
        return list(map(int, numbers))

    return sorted(
        [entry.path.replace("\\", "/") for entry in os.scandir(folder_path) if entry.is_file()],
        key=lambda x: extract_numbers(os.path.basename(x))
    )


def run_original_solver(successor_name, encoder_name, graph_path):
    ref_path = os.path.abspath("src_ref/SAT_HCP/src")
    if ref_path not in sys.path:
        sys.path.insert(0, ref_path)

    # Import original classes
    from Graph import Graph as RefGraph
    from HcpSolver import HcpSolver as RefHcpSolver

    # Map successor
    if successor_name == "unary":
        from successor.successorMethod.Unary import Unary as RefUnary
        successor = RefUnary()
    elif successor_name == "lfsr":
        from successor.successorMethod.LFSR import LFSR as RefLFSR
        successor = RefLFSR()
    else:
        from successor.successorMethod.BinaryAdder import BinaryAdder as RefBinaryAdder
        successor = RefBinaryAdder()

    # Map encoder
    if encoder_name == "binomial":
        from exactlyone.exactlyoneMethod.EOBinomial import EOBinomial as RefEnc
    elif encoder_name == "binary":
        from exactlyone.exactlyoneMethod.EOBinary import EOBinary as RefEnc
    elif encoder_name == "commander":
        from exactlyone.exactlyoneMethod.EOCommander import EOCommander as RefEnc
    elif encoder_name == "product":
        from exactlyone.exactlyoneMethod.EOProduct import EOProduct as RefEnc
    elif encoder_name == "sequential":
        from exactlyone.exactlyoneMethod.EOSequentialEncounter import EOSequentialEncounter as RefEnc
    elif encoder_name == "pblib":
        from exactlyone.exactlyoneMethod.EOPbLib import EOPbLib as RefEnc
    elif encoder_name == "hybrid":
        from exactlyone.exactlyoneMethod.EOHybrid import EOHybrid as RefEnc
    else:
        # Original doesn't support scl or bimander
        # Clean up path and raise
        if ref_path in sys.path:
            sys.path.remove(ref_path)
        raise ValueError(f"Original solver does not support encoder: {encoder_name}")

    encoder = RefEnc()
    hcpSolver = RefHcpSolver(successor, encoder)
    hcpSolver.graph.load_graph_from_file(graph_path)

    cnf = hcpSolver.hcpCnf
    cnf = hcpSolver.successor.build_clauses(cnf, hcpSolver.graph)
    result = hcpSolver.solve(cnf)

    if result["status"] == "SAT":
        hcpSolver.print_result(result["model"], hcpSolver.graph, hcpSolver.successor.getH, result)

    # Clean up sys.path and sys.modules
    if ref_path in sys.path:
        sys.path.remove(ref_path)
    for k in list(sys.modules.keys()):
        if any(prefix in k for prefix in ["Graph", "HcpSolver", "successor.", "exactlyone.", "utils."]):
            del sys.modules[k]

    return result


def run_new_solver(successor_name, encoder_name, graph_path):
    src_path = os.path.abspath("src")
    if src_path not in sys.path:
        sys.path.insert(0, src_path)

    from libs.utils.graph import Graph as NewGraph
    from solver import HcpSolver as NewHcpSolver

    # Map successor
    from libs.models.successor.unary import Unary as NewUnary
    from libs.models.successor.lfsr import LFSR as NewLFSR
    from libs.models.successor.binary_adder import BinaryAdder as NewBinaryAdder
    from libs.models.successor.binary_adder_original import BinaryAdderOriginal as NewBinaryAdderOriginal
    from libs.models.successor.vee import Vee as NewVee

    if successor_name == "unary":
        successor = NewUnary()
    elif successor_name == "lfsr":
        successor = NewLFSR()
    elif successor_name == "binary_adder_original":
        successor = NewBinaryAdderOriginal()
    elif successor_name == "vee":
        successor = NewVee()
    else:
        successor = NewBinaryAdder()

    # Map encoder
    from libs.encodings.cardinality_one.bimander import BimanderEncoder
    from libs.encodings.cardinality_one.binomial import BinomialEncoder
    from libs.encodings.cardinality_one.binary import BinaryEncoder
    from libs.encodings.cardinality_one.commander import CommanderEncoder
    from libs.encodings.cardinality_one.product import ProductEncoder
    from libs.encodings.cardinality_one.sequential import SequentialEncoder
    from libs.encodings.cardinality_one.hybrid import HybridEncoder
    from libs.encodings.cardinality_one.pblib import PbLibEncoder
    from libs.encodings.cardinality_one.scl import NSCEncoding
    from libs.encodings.cardinality_one.heule import HeuleEncoder

    if encoder_name == "bimander":
        encoder = BimanderEncoder()
    elif encoder_name == "binomial":
        encoder = BinomialEncoder()
    elif encoder_name == "binary":
        encoder = BinaryEncoder()
    elif encoder_name == "commander":
        encoder = CommanderEncoder()
    elif encoder_name == "product":
        encoder = ProductEncoder()
    elif encoder_name == "sequential":
        encoder = SequentialEncoder()
    elif encoder_name == "hybrid":
        encoder = HybridEncoder()
    elif encoder_name == "pblib":
        encoder = PbLibEncoder()
    elif encoder_name == "scl":
        encoder = NSCEncoding()
    elif encoder_name == "heule":
        encoder = HeuleEncoder()

    hcpSolver = NewHcpSolver(successor, encoder)
    hcpSolver.graph.load_graph_from_file(graph_path)
    result = hcpSolver.solve()

    if result["status"] == "SAT":
        hcpSolver.print_result(result["model"], hcpSolver.successor.getH, result)

    # Clean up sys.path
    if src_path in sys.path:
        sys.path.remove(src_path)

    return result


def main():
    parser = argparse.ArgumentParser(description="Compare Original vs Refactored HCP SAT Solver.")
    parser.add_argument("-s", "--successor", choices=["unary", "lfsr", "binary_adder", "binary_adder_original", "vee"], default="binary_adder", help="Successor method (default: binary_adder)")
    parser.add_argument("-e", "--encoder", choices=["binomial", "binary", "commander", "product", "sequential", "hybrid", "pblib", "heule"], default="pblib", help="Card-one encoder supported by both solvers")
    parser.add_argument("-d", "--data", type=str, default="src/data/fhcpcs", help="Folder containing graph instances (default: src/data/fhcpcs)")
    parser.add_argument("-n", "--limit", type=int, default=3, help="Limit files to run (default: 3)")

    args = parser.parse_args()
    successor_type = args.successor
    encoder_type = args.encoder
    data_folder = args.data
    limit = args.limit

    if not os.path.exists(data_folder):
        print(f"Data folder {data_folder} not found.")
        sys.exit(1)

    listFiles = get_files_from_folder(data_folder)
    if not listFiles:
        print(f"No files found in {data_folder}.")
        sys.exit(0)

    print(f"Comparing solvers on first {min(limit, len(listFiles))} instances...")
    print(f"Successor: {successor_type.upper()} | Encoder: {encoder_type.upper()}")
    print("-" * 110)

    results = []

    for path in listFiles[:limit]:
        filename = path.split("/")[-1]
        print(f"\nRunning original solver on: {filename}")
        orig_res = run_original_solver(successor_type, encoder_type, path)

        print(f"Running new refactored solver on: {filename}")
        new_res = run_new_solver(successor_type, encoder_type, path)

        results.append({
            "filename": filename,
            "orig_vars": orig_res["nofVariables"],
            "new_vars": new_res["nofVariables"],
            "orig_clauses": orig_res["nofClauses"],
            "new_clauses": new_res["nofClauses"],
            "orig_time": orig_res["time"],
            "new_time": new_res["time"],
            "orig_status": orig_res["status"],
            "new_status": new_res["status"]
        })

    print("\n" + "=" * 45 + " COMPARISON RESULTS " + "=" * 45)
    print(f"{'File Name':<20} | {'Vars (Orig / New)':<20} | {'Clauses (Orig / New)':<24} | {'Time (Orig / New)':<20} | {'Status'}")
    print("-" * 110)
    for r in results:
        vars_str = f"{r['orig_vars']} / {r['new_vars']}"
        clauses_str = f"{r['orig_clauses']} / {r['new_clauses']}"
        time_str = f"{r['orig_time']:.3f}s / {r['new_time']:.3f}s"
        status_str = f"{r['orig_status']} / {r['new_status']}"
        print(f"{r['filename']:<20} | {vars_str:<20} | {clauses_str:<24} | {time_str:<20} | {status_str}")
    print("=" * 110)


if __name__ == "__main__":
    main()

File: src_cpp/compare_hcp.cpp
#include "libs/encodings/cardinality_one/BimanderEncoder.hpp"
#include "libs/encodings/cardinality_one/BinomialEncoder.hpp"
#include "libs/encodings/cardinality_one/BinaryEncoder.hpp"
#include "libs/encodings/cardinality_one/CommanderEncoder.hpp"
#include "libs/encodings/cardinality_one/ProductEncoder.hpp"
#include "libs/encodings/cardinality_one/SequentialEncoder.hpp"
#include "libs/encodings/cardinality_one/HybridEncoder.hpp"
#include "libs/encodings/cardinality_one/NSCEncoding.hpp"
#include "libs/encodings/cardinality_one/PbLibEncoder.hpp"
#include "libs/encodings/cardinality_one/HeuleEncoder.hpp"

#include "libs/models/successor/Unary.hpp"
#include "libs/models/successor/LFSR.hpp"
#include "libs/models/successor/BinaryAdder.hpp"
#include "libs/models/successor/BinaryAdderOriginal.hpp"
#include "libs/models/successor/CRT.hpp"
#include "libs/models/successor/Vee.hpp"
#include "libs/models/successor/IncCRT.hpp"
#include "libs/models/successor/OCRT.hpp"
#include "libs/models/successor/IncOCRT.hpp"
#include "libs/models/successor/OUnary.hpp"
#include "libs/models/successor/PruningCRT.hpp"

#include "libs/utils/Common.hpp"
#include "solver.hpp"

#include <iostream>
#include <memory>
#include <chrono>
#include <iomanip>
#include <fstream>
#include <sstream>
#include <algorithm>

std::unique_ptr<CardinalityOneEncoder> make_encoder(const std::string& type);
std::unique_ptr<SuccessorMethod> make_successor(const std::string& type, CardinalityOneEncoder* encoder, int cycle_len);

struct CompareRow {
    std::string filename;
    int py_vars = 0;
    int cpp_vars = 0;
    int py_clauses = 0;
    int cpp_clauses = 0;
    double py_time = 0.0;
    double cpp_time = 0.0;
    std::string py_status;
    std::string cpp_status;
};

int main(int argc, char* argv[]) {
    std::string successor_type = "binary_adder";
    std::string encoder_type = "pblib";
    std::string data_path = "src/data/fhcpcs";
    int limit = 3;

    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "-s" || arg == "--successor") {
            if (i + 1 < argc) successor_type = argv[++i];
        } else if (arg == "-e" || arg == "--encoder") {
            if (i + 1 < argc) encoder_type = argv[++i];
        } else if (arg == "-d" || arg == "--data") {
            if (i + 1 < argc) data_path = argv[++i];
        } else if (arg == "-n" || arg == "--limit") {
            if (i + 1 < argc) limit = std::stoi(argv[++i]);
        }
    }

    auto listFiles = get_files_from_folder(data_path);
    if (listFiles.empty()) {
        std::cout << "No graph instances found in: " << data_path << std::endl;
        return 0;
    }

    int files_to_run = std::min(limit, (int)listFiles.size());
    std::cout << "Comparing solvers on first " << files_to_run << " instances..." << std::endl;
    std::cout << "Successor: " << successor_type << " | Encoder: " << encoder_type << std::endl;
    std::cout << std::string(110, '-') << std::endl;

    std::vector<CompareRow> rows;

    for (int i = 0; i < files_to_run; ++i) {
        std::string path = listFiles[i];
        std::string filename = fs::path(path).filename().string();

        // 1. Run Python solver as subprocess and redirect log
        std::remove("temp_py_log.txt");
        std::string py_cmd = "python3 src/test_hcp.py -s " + successor_type + " -e " + encoder_type + " -d " + path + " -l temp_py_log.txt > /dev/null 2>&1";
        std::system(py_cmd.c_str());

        CompareRow row;
        row.filename = filename;

        // Parse temp_py_log.txt
        std::ifstream py_log("temp_py_log.txt");
        if (py_log.is_open()) {
            std::string line;
            if (std::getline(py_log, line)) {
                std::stringstream ss(line);
                std::string fname, vars, clauses, status, time;
                ss >> fname >> vars >> clauses >> status >> time;
                row.py_vars = std::stoi(vars);
                row.py_clauses = std::stoi(clauses);
                row.py_status = status;
                row.py_time = std::stod(time);
            }
            py_log.close();
        }
        std::remove("temp_py_log.txt");

        // 2. Run new C++ solver
        auto encoder = make_encoder(encoder_type);
        int cycle_override = -1;
        if (successor_type == "inc_crt" || successor_type == "inc_ocrt") {
            cycle_override = 12;
        }
        auto successor = make_successor(successor_type, encoder.get(), cycle_override);

        Graph graph;
        graph.load_graph_from_file(path);

        SolverResult cpp_res;
        if (successor->is_incremental) {
            IncHcpSolver solver(successor.get(), encoder.get());
            cpp_res = solver.solve_incremental(graph, 0); // Propagator mode
        } else {
            HcpSolver solver(successor.get(), encoder.get());
            cpp_res = solver.solve(graph);
        }

        row.cpp_vars = cpp_res.nofVariables;
        row.cpp_clauses = cpp_res.nofClauses;
        row.cpp_status = cpp_res.status;
        row.cpp_time = cpp_res.time;

        rows.push_back(row);
    }

    std::cout << "\n" << std::string(45, '=') << " COMPARISON RESULTS " << std::string(45, '=') << std::endl;
    std::cout << std::left << std::setw(20) << "File Name" << " | "
              << std::setw(20) << "Vars (Py / C++)" << " | "
              << std::setw(24) << "Clauses (Py / C++)" << " | "
              << std::setw(20) << "Time (Py / C++)" << " | "
              << "Status" << std::endl;
    std::cout << std::string(110, '-') << std::endl;

    for (const auto& r : rows) {
        std::string vars_str = std::to_string(r.py_vars) + " / " + std::to_string(r.cpp_vars);
        std::string clauses_str = std::to_string(r.py_clauses) + " / " + std::to_string(r.cpp_clauses);
        
        std::stringstream time_ss;
        time_ss << std::fixed << std::setprecision(3) << r.py_time << "s / " << r.cpp_time << "s";
        std::string time_str = time_ss.str();

        std::string status_str = r.py_status + " / " + r.cpp_status;

        std::cout << std::left << std::setw(20) << r.filename << " | "
                  << std::setw(20) << vars_str << " | "
                  << std::setw(24) << clauses_str << " | "
                  << std::setw(20) << time_str << " | "
                  << status_str << std::endl;
    }
    std::cout << std::string(110, '=') << std::endl;

    return 0;
}

// Re-using implementations from test_hcp.cpp
std::unique_ptr<CardinalityOneEncoder> make_encoder(const std::string& type) {
    if (type == "bimander") return std::make_unique<BimanderEncoder>();
    if (type == "binomial") return std::make_unique<BinomialEncoder>();
    if (type == "binary") return std::make_unique<BinaryEncoder>();
    if (type == "commander") return std::make_unique<CommanderEncoder>();
    if (type == "product") return std::make_unique<ProductEncoder>();
    if (type == "sequential") return std::make_unique<SequentialEncoder>();
    if (type == "hybrid") return std::make_unique<HybridEncoder>();
    if (type == "pblib") return std::make_unique<PbLibEncoder>();
    if (type == "scl") return std::make_unique<NSCEncoding>();
    if (type == "heule") return std::make_unique<HeuleEncoder>();
    throw std::runtime_error("Unknown encoder: " + type);
}

std::unique_ptr<SuccessorMethod> make_successor(const std::string& type, CardinalityOneEncoder* encoder, int cycle_len) {
    if (type == "unary") return std::make_unique<Unary>(encoder);
    if (type == "lfsr") return std::make_unique<LFSR>(encoder);
    if (type == "crt") return std::make_unique<CRT>(encoder, cycle_len);
    if (type == "inc_crt") return std::make_unique<IncCRT>(encoder, cycle_len);
    if (type == "vee") return std::make_unique<Vee>(encoder);
    if (type == "ocrt") return std::make_unique<OCRT>(nullptr, encoder, cycle_len);
    if (type == "inc_ocrt") return std::make_unique<IncOCRT>(nullptr, encoder, cycle_len);
    if (type == "ounary") return std::make_unique<OUnary>(nullptr);
    if (type == "pruning_crt") return std::make_unique<PruningCRT>(encoder, cycle_len);
    if (type == "binary_adder_original") return std::make_unique<BinaryAdderOriginal>(encoder);
    if (type == "binary_adder") return std::make_unique<BinaryAdder>(encoder);
    throw std::runtime_error("Unknown successor: " + type);
}
